### ISSUE TITLE
Have changelog inside the repo. 

### ISSUE THREAD

**Issue by shirishag75** [mars-sim/mars-sim#25]
See for e.g.  https://github.com/egoist/changelog.md . It would be nice to have changelog from within the repo. itself. It could also be gunzipped for maximum compressibility.

**Comment by shirishag75** [comment]
if you are using git on linux and having tagged releases, getting diffed changelog is pretty easy, for e.g. - 

For e.g. a repo. I cloned today itself only - 

```
┌─[shirish@debian] - [~/games/gzip] - [10113]
└─[$] git-info | grep remote.origin.url                                                                                              
remote.origin.url=git://git.savannah.gnu.org/gzip.git
```

and saw what the changes were between tagged releases -

```
[$] git shortlog v1.6...v1.7 

     1	Jason Leschnik (1):
     2	      doc: correct a diagnostic in man page to match actual
     3	
     4	Jim Meyering (23):
     5	      maint: post-release administrivia
     6	      maint: update gnulib to latest and adapt streamsavedir usage
     7	      maint: update copyright year range in gzip.texi
     8	      maint: udpate all copyright notices via "make update-copyright"
     9	      gnulib+bootstrap: update to latest
    10	      maint: avoid false positive match in check for double semicolon
    11	      maint: enable more syntax checks
    12	      maint: move new NEWS entry into block for upcoming release
    13	      maint: update copyright year ranges to include 2015; update gnulib
    14	      gnulib: update to latest
    15	      maint: ensure that --version's copyright date is current
    16	      maint: remove dead code
    17	      build: avoid -Wshift-negative-value warning
    18	      maint: adjust copyright notices in *.in files to be consistent
    19	      maint: avoid three warnings from the very latest gcc-built-from-git
    20	      maint: update copyright year, bootstrap, init.sh
    21	      maint: dist-check.mk: remove .deps dirs before comparing
    22	      tests: fix "make check" failure on AIX 7.1
    23	      tests: port to systems for which ":" is not the PATH separator
    24	      maint: don't ignore gitlog-to-changelog failure
    25	      gnulib: update to latest
    26	      gzip: also honor GZIP=--rsyncable
    27	      version 1.7
    28	
    29	Paul Eggert (35):
    30	      maint: port to platforms lacking SIGPIPE
    31	      doc: zgrep exit status, unsupported options
    32	      tests: zgrep-context assumes grep knows context
    33	      gzip: port util.c to Compaq C V6.5-303
    34	      zgrep: usage should say which grep options are not supported
    35	      tests: zgrep-signal race condition fix
    36	      Fix spelling typo in newly-added comment.
    37	      znew: avoid denial-of-service issue
    38	      gzip: fix permissions issue on Solaris-like systems
    39	      zless: improve gzip failure checking, and port to new -V format
    40	      maint: update .gitignore files
    41	      doc: use UTF-8 in the manual
    42	      zgrep: exit with status 0 if a file matches and there's no trouble
    43	      gzip: fix --suffix=z bug (Bug#18239)
    44	      tests: use local dir for output
    45	      gzip: adjust -v output when -k is also specified
    46	      maint: adjust to recent gnulib
    47	      gzip: make the GZIP env var obsolescent
    48	      fsync output file before closing
    49	      gzip: fdatasync output dir before unlinking
    50	      gzip: use constants, not fileno
    51	      gzip: new option --synchronous
    52	      misc: update --version copyright
    53	      gzip: remove --__bindir
    54	      gzip: port zgrep to Solaris 11.2
    55	      gzip: minor zgrep cleanup
    56	      doc: minor --rsyncable doc fixes
    57	      gzip: pacify clang
    58	      gzip: port to AIX 7.1 + xlc V12.1
    59	      zgrep: with -f SPECIAL, read SPECIAL just once
    60	      Port to Solaris 11 /bin/sh
    61	      Fix typo in previous patch
    62	      Port to Alpine Linux which uses Busybox
    63	      * tests/init.sh: Sync from Gnulib.
    64	      Port to NetBSD 7.0
    65	
    66	Rusty Russell (1):
    67	      gzip: support the --rsyncable option
    68	

```

And from 1.7 to 1.8 

```
[$] git shortlog v1.7...v1.8

     1	      Jim Meyering (3):
     2	      maint: post-release administrivia
     3	      gnulib: update to latest
     4	      version 1.8
     5	
     6	Paul Eggert (3):
     7	      Port to Oracle Solaris Studio 12.4
     8	      gzip: fix bug with -l output to pipes
     9	      gzip: simplify by closing ourselves
    10

```

One of the reasons why git is loved so much :)

**Comment by shirishag75** [comment]
will do once the migration is done and I have some idea of what the repo. size will be.

**Comment by mokun** [comment]
sure! 

Do you mean the changelog FOR the repo itself, right ? 

When it comes to the changelog for mars-sim, there's a changelog for each preview version of 3.1.0 for mars-sim to document the updates. mars-sim has an internal version_history.html under mars-sim-ui's /resources/docs/help/ 

Can you make one for the repo ? 

Thx much!

**Comment by shirishag75** [comment]
Usually in GNU/Linux you have two different changelogs, the one which I meant was for the repo. and another one which is called as NEWS.gz which you called as version_history.html . To illustrate the differences I will share content from an application/package called guake which is a drop-down terminal emulator (reminds almost everybody of Quake the game, hence the name ) -

```
┌─[shirish@debian] - [/usr/share/doc/guake] - [10140]
└─[$] ls                                                                                                                           
changelog.Debian.gz   changelog.gz  copyright  NEWS.gz  README.rst.gz
```

Now see what changelog.gz tells and what NEWS.gz tells - 

```
[$] zless changelog.gz                                                                                                              
(news continues on the NEWS file)

2008-11-25  Gabriel Falcão <gabriel@nacaolivre.org>

	* src/guake.py: Applying Licio's (liciofernando@gmail.com) patch for
	dual head monitors.

2008-08-09  Gabriel Falcão <gabriel@nacaolivre.org>

	* po/pt_BR.po: Fixing a little semantic error.
	* src/guake.py: on delete_tab(), pre-adding a new tab when all tabs get
	closed, it improves the speed of new guake show request
	* src/guake.py: Guake openning new terminals in the same directory than
	current openned tab.

2008-08-07  Gabriel Falcão <gabriel@nacaolivre.org>

	* TODO: Somethings turn obselete since new features had been implemented.
	* configure.ac: Guake version 0.4.0
	* src/guake.py: Avoinding gdk device mods mistakes then handling the
	resizer.

2008-08-05  Gabriel Falcão <gabriel@nacaolivre.org>

	* src/guake.py: Now guake can be resized with a resizer :)
	* data/guake.glade: Adding a GtkHPaned as resizer for guake
	* data/guake.schemas: Adding two new boolean keys: show_toolbar and show_resizer
	* data/guake.glade: toolbar (tabs) is no more affected by window.show_all()
	* src/guake.py: Guake are listening to GConf! (only show_resizer and
	show_toolbar stuff for now)
	* data/guake.schemas: Adding a new boolean key:	toolbar_visible_in_fullscreen
	* src/guake.py: Guake checking if the tabs must be visible when in
	fullscreen mode.
```

As can be seen these are commit messages bundled by date/compressed - 

```
┌─[shirish@debian] - [/usr/share/doc/guake] - [10144]
└─[$] zless NEWS.gz

Version 0.8.8
=============

Thank to these contribution from Guake users, I am happy to announce a new minor fix release of
Guake.

Features:
* Close a tab with the middle button of the mouse

Bug Fixes:
- Fix error when toggle key was disabled
- Update change news
- Uppercase pallete name
- Fix pylint errors
- Convert README badge to SVG
- Update Japanese translation
- update Russian translation
- updated CS translation
- Update zh_CN translation


Version 0.8.7
=============

Do not forget to update the software version
```

**Comment by shirishag75** [comment]
some packages/applications also have TODO.gz and things like that so any programmers/developers if they like the program they know what would be the things upstream is aware of and need/would like help with.

**Comment by mokun** [comment]
sure! as soon as we switch to using git, we can start experimenting with adding some of these files. 

NEWS and changelog tracked or made by git may not be realistic in case of mars-sim, where hundreds of changes are made between each release. 

In terms of TODO, we'll need to improve our project management flow to make use of it for the next release.

**Comment by shirishag75** [comment]
have you moved to using git or still using svn ?

**Comment by larsnaesbye** [comment]
We have moved to git :)

**Comment by mokun** [comment]
It's been on git since Jan/Feb 2017.

**Comment by shirishag75** [comment]
Any update on the above - 

One of the examples that you could use as a reference point could perhaps be - 

https://github.com/rg3/youtube-dl/blob/master/ChangeLog

or 

https://github.com/qbittorrent/qBittorrent/blob/master/Changelog

There are many number of ways to achieve this, some of the good ones shared are at  https://stackoverflow.com/questions/3523534/good-ways-to-manage-a-changelog-using-git 

just use one or the other whichever system you feel would be good. You would also have to think about how you are going to number/tag your releases, two of the most common ways are - 

major.minorversion tarball so it could be 0.0.1 mars-sim.tar.gz OR 

date-timestamp version like youtube-dl does, for them that works as they are constantly being updated  even in our distributions every month or two. See - 
```

$ apt-cache policy youtube-dl
youtube-dl:
  Installed: 2018.04.25-1
  Candidate: 2018.04.25-1
  Version table:
 *** 2018.04.25-1 900
        900 http://cdn-fastly.deb.debian.org/debian buster/main amd64 Packages
        100 http://cdn-fastly.deb.debian.org/debian unstable/main amd64 Packages
        100 /var/lib/dpkg/status
```

I would suggest the former method but that's upto the devs. who are actually going to do the releases.

**Comment by mokun** [comment]
Looks like there's a lot of favors of how this CHANGELOG.md file may look like and we just have to pick one that works for us.

For me, since I usually write in details whenever I commit the change in git, I'll be interested capturing this note and may be simplifying it in such a way that it will automatically append the content of the change into the file CHANGELOG.md  

For instance, I have the following commit message for build 4256 : 

----------------------------------------------------------------------------
Add "noaudio" argument

r4256
2018-02-18

Note: this patch is issued in response to a feature request at
#7

# NEW
1. Add new argument "noaudio" if user prefers no background music and
   sound effect.
   - Add noaudio option in MarsProjectFX, MarsProjectStarter
   - Add boolean isSoundDisabled, isSoundDisabled(), disableSound()
     in AudioPlayer
   - Add isSoundDisabled, setSoundDisabled(), isSoundDisabled()
     in MainMenu.
   - Update "noaudio" option in README and CHANGELOG.
2. Add two new eclipse launchers
   - "MarsProjectFX -new -noaudio.launch"
   - "MarsProjectFX -load -noaudio.launch"

# CHANGE
1. Change musicMuteBox and soundEffectMuteBox to static in MainScene.
2. Modify updateTimeLabels() in MainScene to check if sound is
   disabled.
3. Call disableSound() in createSoundPopup if sound is disabled.
4. Update year from 2017 to 2018 and GNU license from version 2 to 3
   in copyright, changes.xml, MenuApp.

**Comment by shirishag75** [comment]
no, you should not have long commit messages, have a changelog.md in which whenever you or any other member of the team either make a change to commits or make a release the changelog.md should take the commit messages and make the changelog. 

If you want to do it manually, then do it by making a changelog.md and pushing the changes therein instead of having a large commit message. What distributions do, at least I know about Debian is it would take the changelog and make changelog.gz (compress) and stove it at /usr/share/doc/changelog.gz . That would be the packagers/maintainers work . 

All the other changes could be in another file called changelog.historical which could be one-time gitlog dump, something like - 

`git log --since=3/14/2000 --until=3/14/2017 --no-merges --format=%B`

**Comment by akademy** [comment]
What do you imagine an automatically generated changlelog from the git revisions will be used for? It doesn't help users because it'll be far to technical, and if a programmer needs one they can just generate it directly from git.

What is useful is a list of changes between versions that the user would be interested in. (e.g. "Improved speed" rather than "replaced my_func() with faster algorithm" )

**Comment by shirishag75** [comment]
that's true, so please go ahead with that.

**Comment by shirishag75** [comment]
one way you could do is squashing commits, let's say you do 4-5 commits for some feature or function and then at the end just squash them altogether. 

https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git 

it does give some various ways you could do that, maybe there may be one which is appealing to you.

**Comment by mokun** [comment]
Sure. I like things evolving and don't want it to stay the same. So I look forward to finding a good changelog.md format to adopt.

btw, we've been using travis ci for a while.

If you click on the green check mark found on each commit at the Code page, it will take you to the full log of how well the latest build is compiled with the help of travis ci.

see https://travis-ci.org/mars-sim/mars-sim/builds/389918077?utm_source=github_status&utm_medium=notification

**Comment by mokun** [comment]
I came across this neat [article](https://keepachangelog.com/en/1.0.0/) on changelog. The author even has a a changelog [podcast](https://changelog.com/podcast/127).
